src/renuo_upload.ts

```diff
@@ -0,0 +1,139 @@
+  }
+
+  private getShortName(cleanName:string):string {
+    return cleanName.replace(/\.[^/.]+$/, '');
+  }
+
+  private getPublicUrl(cleanName:string):string {
+    return `${this.fileUrlPath}#{cleanName}`;
+  }
+
+  private getFilePath(cleanName:string):string {
+    return `${this.filePrefix}#{cleanName}`;
+  }
+
+  private getUploadInfoAndSignature() {
+    jQuery.ajax({
+      type: 'POST',
+      url: this.signingUrl,
+      data: {
+        api_key: this.apiKey,
+      },
+      dataType: 'json'
+    }).done((responseJson) => {
+      this.dropzoneOptions.url = responseJson.url;
+      this.dropzoneOptions.params = {};
+      jQuery.each(responseJson.data, (k:string, v:string) => {
+        this.dropzoneOptions.params[k.replace(/_/g, '-')] = v;
+      });
+      this.filePrefix = responseJson.file_prefix;
+      this.fileUrlPath = responseJson.file_url_path;
+    }).fail(() => {
+      throw new Error('Failed to get credential for upload.');
+    });
+  }
+
+  private buildResult(file):RenuoUploadResult {
+    const cleanFilename:string = this.cleanFilename(file.name);
+    return {
+      orginalName: file.name,
+      cleanName: cleanFilename,
+      name: this.getShortName(cleanFilename),
+      extension: this.getExtension(file.name),
+      size: file.size,
+      publicUrl: this.getPublicUrl(cleanFilename),
+      filePath: this.getFilePath(cleanFilename)
+    };
+  }
+
+  private checkAdaptParams() {
+    this.checkElement();
+    this.adaptCallback();
+    this.adaptOptions();
+  }
+
+  private checkRequirements() {
+    if (!jQuery) throw new Error('RenuoUpload needs jQuery.');
+    if (!Dropzone) throw new Error('RenuoUpload needs Dropzone.');
+  }
+
+  private checkElement() {
+    if (!this.element) throw new Error('Element is not defined');
+    if (!this.element[0].nodeType) this.element = this.element[0];
+    if (!this.element.nodeType) throw new Error('Element is not a valid element');
+  }
+
+  private adaptOptions() {
+    if (!this.dropzoneOptions) throw new Error('DropzoneOptions is not defined');
+    if (!this.dropzoneOptions.acceptedFiles) {
+      throw new Error('DropzoneOptions.acceptedFiles is not defined');
+    }
+    if (typeof this.dropzoneOptions.acceptedFiles !== 'string') {
+      throw new Error('DropzoneOptions.acceptedFiles is not a string');
+    }
+    if (!this.dropzoneOptions.parallelUploads) this.dropzoneOptions.parallelUploads = 25;
+    if (!this.dropzoneOptions.renameFilename) this.dropzoneOptions.renameFilename = this.cleanFilename;
+  }
+
+  private adaptCallback() {
+    if (typeof this.callback !== 'function')
+      this.callback = this.defaultCallback;
+  }
+
+  private defaultCallback(result:RenuoUploadResult) {
+    if (jQuery(this.element).parents('form').length) {
+      name = result.name;
+      //TODO discuss with Lukas about upload of two times the same file
+      delete result.name;
+      //TODO check if necessary, unexpected behavior, good point from Y
```

Comment:
Yes, would be a good moment to solve them.

Problem 1:  Second file uploaded with the same name will override the first
Problem 2: The delete is fishy (we remove the name attribute from the object and iterate over all other attributes)

Problem 1: I think the behaviour is okay, but we have always a "dead" file on the server. Do we need to take action?
Problem 2: We could rewrite it with an if, would probably be more readable, or any other suggestion?